### ISSUE TITLE
Problems when reading special procfs file

### ISSUE THREAD

**Issue by hexchain** [dart-lang/sdk#25596]
OS: 4.4.0-4-ARCH

Reading special procfs file (e.g. `/proc/$PID/smaps`) as chunk does not always return a full chunk before reaching EOF.

Currently Dart determines EOF by trying to read BLOCK_SIZE of bytes and checking if the length of returned bytes equals to BLOCK_SIZE, which may fail in such special circumstances. Relevant code is [here](https://github.com/dart-lang/sdk/blob/master/sdk/lib/io/file_impl.dart#L483) and [here](https://github.com/dart-lang/sdk/blob/master/sdk/lib/io/file_impl.dart#L450), maybe more.

And here's a simple test in C:
```c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#define BLOCK_SIZE 64 * 1024

int main(void)
{
    int fd = open("/proc/918/smaps", O_RDONLY); // change the PID part accordingly
    int r = 0, t = 0;
    char buf[BLOCK_SIZE];
    do {
        r = read(fd, buf, BLOCK_SIZE);
        t += r;
        printf("r = %d\n", r); // r is not always BLOCK_SIZE
    } while(r > 0);  // Notice this expression
    printf("t = %d\n", t);
    return 0;
}
```

**Comment by sgjesse** [comment]
Thanks for the report. Proposed fix in https://codereview.chromium.org/1651023003.

**Comment by sgjesse** [comment]
Will be released in 1.14.2.
